lms/djangoapps/support/views/program_enrollments.py

```diff
@@ -98,3 +108,185 @@ def _validate_and_link(program_uuid_string, linkage_text):
+                'org_keys': self._get_org_keys_with_IdP_provider()
+            }
+        )
+
+    def _get_org_keys_with_IdP_provider(self):
+        """
+        From our Third_party_auth models, return a list
+        of organizations whose SAMLProviders are active and configured
+        """
+        orgs_with_saml = []
+        saml_providers = SAMLProviderConfig.objects.current_set().filter(
+            enabled=True,
+            organization__isnull=False
+        ).select_related('organization')
+
+        for saml_provider in saml_providers:
+            orgs_with_saml.append(saml_provider.organization.short_name)
+
+        return orgs_with_saml
+
+    def _get_account_info_by_external_key(self, org_key, external_user_key):
+        """
+        Provided the external_user_key and the organization the external_key
+        belongs, return edx account info and program_enrollments_info.
+        If we cannot find relevant info, return empty object and error
+        """
+        return {}, ''
+
+    def _get_account_info(self, username_or_email):
+        """
+        Provided the edx account username or email, return edx account info
+        and program_enrollments_info. If we cannot identify the user, return
+        empty object and error.
+        """
+        user_info = {}
+        external_key = None
+        try:
+            user = User.objects.get(Q(username=username_or_email) | Q(email=username_or_email))
+            user_info['username'] = user.username
+            user_info['email'] = user.email
+            try:
+                user_social_auth = UserSocialAuth.objects.get(user=user)
+                slug, external_key = user_social_auth.uid.split(':', 1)
+                user_info['external_user_key'] = external_key
+                user_info['SSO'] = {
+                    'uid': user_social_auth.uid,
+                    'provider': user_social_auth.provider
+                }
+            except UserSocialAuth.DoesNotExist:
+                pass
+
+            enrollments = self._get_enrollments(user=user)
+            result = {'user': user_info}
+            if enrollments:
+                result['enrollments'] = enrollments
+
+            return result, ''
+        except User.DoesNotExist:
+            return {}, 'Could not find edx account with {}'.format(username_or_email)
+
+    def _get_enrollments(self, user=None, external_user_key=None):
+        """
+        With the user or external_user_key passed in,
+        return a dictionary with corresponding ProgramEnrollments and ProgramCourseEnrollments
+        all serialized for view
+        """
+        program_enrollments = fetch_program_enrollments_by_student(
+            user=user,
+            external_user_key=external_user_key
+        )
+        enrollments_by_program_uuid = {}
+
+        program_course_enrollments = get_program_course_enrollments_by_program_enrollments(
+            program_enrollments
+        )
```

Comment:
I think `fetch_program_enrollments_by_student` returns a QuerySet, to which we can add a prefetch_related even after filtering, no? This seems a little excessively over-adherent to "no fetching logic except in the reading module"

Reply:
Feedback taken. Thank you for the suggestion